pkg/kubelet/rocket/rocket.go

```diff
@@ -0,0 +1,453 @@
+	for _, line := range output {
+		tuples := strings.Split(strings.TrimSpace(line), "\t")
+		if len(tuples) != 3 {
+			continue
+		}
+		result[tuples[0]] = tuples[2]
+	}
+	return result, nil
+}
+
+// getPodStatus fills the status of the given pod. Especially, it will
+// fill the status for each container in the pod.
+func (r *Runtime) getPodStatus(pod *api.Pod) error {
+	// TODO(yifan) Cache this.
+	podStates, err := r.getPodsState()
+	if err != nil {
+		return err
+	}
+
+	rktID, found := pod.Annotations[rocketIDKey]
+	if !found {
+		// TODO(yifan): Maybe we should panic here...
+		return fmt.Errorf("rocket: cannot find rocket pod: %v, this is impossible", pod)
+	}
+
+	state, found := podStates[rktID]
+	if !found {
+		return fmt.Errorf("rocket: cannot find the state for pod: %q, rocket ID: %q", pod.Name, rktID)
+	}
+
+	// For now just make every container's state as same as the pod.
+	pod.Status.Info = make(map[string]api.ContainerStatus)
+	for _, container := range pod.Spec.Containers {
+		// TODO(yifan): Pull out creationg, starting time.
+		switch state {
+		case podStateRunning:
+			pod.Status.Info[container.Name] = api.ContainerStatus{
+				State: api.ContainerState{
+					Running: &api.ContainerStateRunning{},
+				},
+			}
+		case podStateEmbryo, podStatePreparing, podStatePrepared:
+			pod.Status.Info[container.Name] = api.ContainerStatus{
+				State: api.ContainerState{
+					Waiting: &api.ContainerStateWaiting{},
+				},
+			}
+		case podStateExited, podStateDeleting, podStateGone:
+			pod.Status.Info[container.Name] = api.ContainerStatus{
+				State: api.ContainerState{
+					Termination: &api.ContainerStateTerminated{},
+				},
+			}
+		default:
+			return fmt.Errorf("rocket: unexpected state: %q", state)
+		}
+	}
+	return nil
+}
+
+// makePod constructs the pod by reading information from the given unit file
+// and from rocket APIs.
+func (r *Runtime) makePod(unitName string) (*api.Pod, error) {
+	f, err := os.Open(path.Join(systemdServiceDir, unitName))
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	var pod api.Pod
+	opts, err := unit.Deserialize(f)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, opt := range opts {
+		if opt.Section != unitKubernetesSection {
+			continue
+		}
+		if opt.Name == unitPodName {
+			// NOTE: In fact we unmarshal from a serialized
+			// api.BoundPod type here.
+			err = json.Unmarshal([]byte(opt.Value), &pod)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	if err = r.getPodStatus(&pod); err != nil {
+		glog.Errorf("Error getting pod status of pod %q.", pod.Name)
+		return nil, err
+	}
+	return &pod, nil
+}
+
+// makePodServiceFileName constructs the unit file name for the given pod.
+// TODO(yifan), when BoundPod and Pod are combined, we can change the input to
+// take just a pod.
+func makePodServiceFileName(podName, podNamespace string) string {
+	return fmt.Sprintf("%s_%s_%s.service", kubernetesUnitPrefix, podName, podNamespace)
```

Comment:
@jonboulle Do you mean the string might contain illegal chars like "#$@#" ? Or do you mean there is a concern for duplicated names?

A safer way is to use UID of the pod, but I just made this to be human readable, so friendly for debugging.

I think it's fine now, let me add a TODO.